python/hsml/core/explicit_provenance.py

```diff
@@ -0,0 +1,338 @@
+                links.faulty.extend(new_links.faulty)
+                links.accessible.extend(new_links.accessible)
+                links.inaccessible.extend(new_links.inaccessible)
+                links.deleted.extend(new_links.deleted)
+        return links
+
+    @staticmethod
+    def __parse_training_datasets(links_json: dict, artifacts: Set[str]):
+        from hsfs.core import explicit_provenance as hsfs_explicit_provenance
+
+        links = Links()
+        for link_json in links_json:
+            if link_json["node"]["artifact_type"] in artifacts:
+                if link_json["node"].get("exception_cause") is not None:
+                    links._faulty.append(
+                        hsfs_explicit_provenance.Artifact.from_response_json(
+                            link_json["node"]
+                        )
+                    )
+                elif bool(link_json["node"]["accessible"]):
+                    links.accessible.append(
+                        hsfs_explicit_provenance.Artifact.from_response_json(
+                            link_json["node"]
+                        )
+                    )
+                elif bool(link_json["node"]["deleted"]):
+                    links.deleted.append(
+                        hsfs_explicit_provenance.Artifact.from_response_json(
+                            link_json["node"]
+                        )
+                    )
+                else:
+                    links.inaccessible.append(
+                        hsfs_explicit_provenance.Artifact.from_response_json(
+                            link_json["node"]
+                        )
+                    )
+        return links
+
+    @staticmethod
+    def from_response_json(json_dict: dict, direction: Direction, artifact: Type):
+        """Parse explicit links from json response. There are three types of
+        Links: UpstreamFeatureGroups, DownstreamFeatureGroups, DownstreamFeatureViews
+
+        # Arguments
+            links_json: json response from the explicit provenance endpoint
+            direction: subset of links to parse - UPSTREAM/DOWNSTREAM
+            type: subset of links to parse - FEATURE_VIEW/TRAINING_DATASET/MODEL
+
+        # Returns
+            A ProvenanceLink object for the selected parse type.
+        """
+
+        import importlib.util
+
+        if not importlib.util.find_spec("hsfs"):
+            raise ValueError(
+                "hsfs is not installed in the environment - cannot parse feature store artifacts"
+            )
+        if not importlib.util.find_spec("hopsworks"):
+            raise ValueError(
+                "hopsworks is not installed in the environment - cannot switch from hsml connection to hsfs connection"
+            )
+
+        # make sure the hsfs connection is initialized so that the feature view/training dataset can actually be used after being returned
+        import hopsworks
+
+        hopsworks._connected_project.get_feature_store()
```

Comment:
Can we add a check here that _connected_project is not None? https://github.com/logicalclocks/hopsworks-api/blob/88488da69aa40c3dd6b1ef577e3c4222b093cbe5/python/hopsworks/__init__.py#L41

Reply:
Because if a user will not have called hopsworks.login(), this value will be None.

Reply:
I added this:
`if not hopsworks._connected_project:
                raise Exception("hopsworks connection is not initialized - use hopsworks.login to connect if you want the ability to use provenance with connections between hsfs and hsml")`
Hope the error message is not too long. But it should be informative